pkg/store/cache/caching_bucket.go

```diff
@@ -132,35 +221,221 @@ func (cb *CachingBucket) ReaderWithExpectedErrs(expectedFunc objstore.IsOpFailur
 	return cb.WithExpectedErrs(expectedFunc)
 }
 
-var chunksMatcher = regexp.MustCompile(`^.*/chunks/\d+$`)
+func (cb *CachingBucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	cfgName, cache, cfg := cb.findCacheConfig(cb.operationConfigs[operationIter], dir)
+	if cfg == nil {
+		return cb.Bucket.Iter(ctx, dir, f)
+	}
+	iterCfg := cfg.(iterConfig)
+
+	cb.operationRequests.WithLabelValues(operationIter, cfgName).Inc()
+
+	key := cachingKeyIter(dir)
+
+	data := cache.Fetch(ctx, []string{key})
+	if data[key] != nil {
+		list, err := decodeIterResult(data[key])
```

Comment:
Why not tie-ing this method to `iterConfig` itself? Easier to version encodings (:

Reply:
Good idea, will do that.

Reply:
Added `IterCodec` interface with `JsonIterCodec` implementation, and removed snappy compression from it. For Thanos, compression wouldn't help much since there is little repetition in block names in root ("") dir.